### PR TITLE
Install cluster monitoring with system account

### PR DESCRIPTION
**Pull request by orangedeng**
**Problem:**
We are using cluster creator id as cluster monitoring creator. There
will be issue if the cluster creator is disabled or deleted.

**Solution:**
We should use system account to install cluster monitoring.

Related issue: https://github.com/rancher/rancher/issues/18787

### CODE REVIEW

File: pkg/controllers/user/helm/controller.go [hunk 1]
```diff
@@ -373,3 +396,16 @@ func (l *Lifecycle) writeKubeConfig(obj *v3.App, tempDir string, remove bool) (s
...
+
+func (l *Lifecycle) isSystemProject(app *v3.App) (bool, error) {
+	project, err := l.projectLister.Get(l.ClusterName, app.Namespace)
+	if err != nil {
+		return false, err
+	}
+	return labels.Set(systemProjectLabel).AsSelector().Matches(labels.Set(project.Labels)), nil
+}
+
+func (l *Lifecycle) isSystemComponent(app *v3.App) bool {
```
Comment (leodotcloud) [inline pkg/controllers/user/helm/controller.go]: I understand what you are trying to do here, but the enclosing function name is quite confusing? Can I call this function with a different app to figure out if it's a system component? I think you should have some kind of annotation or a label on the app to make it a system component. If not, just pick a different function name, something like `isSystemMonitoringApp`.

File: pkg/monitoring/monitoring.go [hunk 2]
```diff
@@ -24,6 +24,7 @@ const (
 const (
 	cattleNamespaceName                    = "cattle-prometheus"
 	cattleCreatorIDAnnotationKey           = "field.cattle.io/creatorId"
+	EnablerAnno                            = "field.cattle.io/monitoringEnabler"
```
Comment (leodotcloud) [inline pkg/monitoring/monitoring.go]: Do we need another annotation to be introduced? If it's a system component, launched in the system project, set the creatorId to system user.

File: pkg/controllers/user/helm/controller.go [hunk 3]
```diff
@@ -346,15 +354,30 @@ func (l *Lifecycle) createAppRevision(obj *v3.App, template, notes string, faile
...
-	} else if errors.IsNotFound(err) && remove {
-		token, err = l.SystemAccountManager.GetOrCreateProjectSystemToken(obj.Namespace)
-	} else if err == nil {
-		token, err = l.UserManager.EnsureToken(helmTokenPrefix+user.Name, description, user.Name)
+	}
+
+	if l.isSystemComponent(obj) {
+		if isSystemProject { // If the app is being deployed into system project, we will create a cluster admin token for it.
+			token, err = l.SystemAccountManager.GetOrCreateClusterSystemToken(l.ClusterName)
+		} else { // If the app is being deployed into user project, we will create a project member token for it.
```
Comment (mrajashree) [inline pkg/controllers/user/helm/controller.go]: `isSystemComponent` will always check if the app being installed is `cluster-monitoring` and in the namespace `cattle-prometheus`. This ns will always be in system project, I'm not sure I get why we need to check `isSystemProject` again?

**Comment by orangedeng** [comment]
Updated, get feedback from Alena and Craig.
/cc @alena1108 @cjellick

**Comment by mrajashree** [comment]
This PR seems to resolve this issue too https://github.com/rancher/rancher/issues/19596

**Comment by leodotcloud** [comment]
@orangedeng can you please rebase, there is a conflict.

File: pkg/controllers/management/globaldns/globaldnsprovider_catalog_launcher.go [hunk 4]
```diff
@@ -172,9 +174,14 @@ func (n *ProviderCatalogLauncher) createUpdateExternalDNSApp(obj *v3.GlobalDNSPr
...
 
+		creator, err := n.systemAccountManager.GetSystemUser(localClusterName)
+		if err != nil {
+			return nil, err
+		}
+
 		toCreate := pv3.App{
 			ObjectMeta: metav1.ObjectMeta{
-				Annotations:     CopyCreatorID(nil, obj.Annotations),
+				Annotations:     map[string]string{cattleCreatorIDAnnotationKey: creator.Name},
```
Comment (leodotcloud) [inline pkg/controllers/management/globaldns/globaldnsprovider_catalog_launcher.go]: Do we have to copy any annotations from `obj`? `CopyCreatorID` function kind intended to do that I guess.

File: pkg/controllers/user/monitoring/clusterHandler.go [hunk 5]
```diff
@@ -362,10 +363,15 @@ func (ch *clusterHandler) deployApp(appName, appTargetNamespace string, appProje
...
+	creator, err := ch.app.systemAccountManager.GetSystemUser(ch.clusterName)
+	if err != nil {
+		return nil, err
+	}
+
 	appCatalogID := settings.SystemMonitoringCatalogID.Get()
 	app := &v3.App{
 		ObjectMeta: metav1.ObjectMeta{
-			Annotations: monitoring.CopyCreatorID(nil, cluster.Annotations),
+			Annotations: map[string]string{creatorIDAnno: creator.Name},
```
Comment (leodotcloud) [inline pkg/controllers/user/monitoring/clusterHandler.go]: Same here. Do we have to copy annotations from `obj`?

File: pkg/controllers/management/globaldns/globaldnsprovider_catalog_launcher.go (hunk 4 above)
Comment (orangedeng) [reply pkg/controllers/management/globaldns/globaldnsprovider_catalog_launcher.go]: We just copy the creator id from obj before.

File: pkg/controllers/user/monitoring/clusterHandler.go (hunk 5 above)
Comment (orangedeng) [reply pkg/controllers/user/monitoring/clusterHandler.go]: ditto

**Comment by orangedeng** [comment]
Updated

**Comment by orangedeng** [comment]
Rebased and added test.

File: pkg/controllers/user/alert/deployer/deployer.go [hunk 6]
```diff
@@ -120,7 +123,7 @@ func (d *Deployer) sync() error {
 		return err
 	}
 
-	systemProjectCreator := systemProject.Annotations[creatorIDAnn]
+	// systemProjectCreator := systemProject.Annotations[creatorIDAnn]
```
Comment (mrajashree) [inline pkg/controllers/user/alert/deployer/deployer.go]: Can we remove this comment?

File: tests/integration/test_system_app_creator.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    provider_name = random_str()
+    access = random_str()
+    secret = random_str()
+    globaldns_provider = \
+        client.create_global_dns_provider(
+                                         name=provider_name,
+                                         rootDomain="example.com",
+                                         route53ProviderConfig={
+                                             'accessKey': access,
+                                             'secretKey': secret})
```
Comment (mrajashree) [inline tests/integration/test_system_app_creator.py]: We can use remove_resource so this provider resource gets deleted after test runs

File: tests/integration/test_system_app_creator.py [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    app = wait_for_system_app(
+        admin_system_pc.client,
+        "systemapp-"+globaldns_provider.name)
+    # the creator id of system app won't be listed in api
+    users = client.list_user(id=app.creatorId)
+    assert len(users) == 0
```
Comment (mrajashree) [inline tests/integration/test_system_app_creator.py]: I get we can't assert that the system account of local cluster is the creator of the app, but we can assert that the creator of this gdns_provider (admin in this case) is not the creator of the app.

**Comment by orangedeng** [comment]
Updated. @mrajashree

File: tests/integration/test_system_app_creator.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    access = random_str()
+    secret = random_str()
+    globaldns_provider = \
+        client.create_global_dns_provider(
+                                         name=provider_name,
+                                         rootDomain="example.com",
+                                         route53ProviderConfig={
+                                             'accessKey': access,
+                                             'secretKey': secret})
+    app = wait_for_system_app(
```
Comment (mrajashree) [inline tests/integration/test_system_app_creator.py]: I think we should use `remove_resource` immediately after the call to resource creation, the wait_for_system_app could fail before remove_resource is applied.

**Comment by mrajashree** [comment]
Just one comment about the test @orangedeng 
apart from that LGTM

**Comment by orangedeng** [comment]
Updated

**Pull request merged**